apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -21,13 +28,136 @@ import { getQuestionCourse } from '../../lib/question-variant.js';
 import * as questionServers from '../../question-servers/index.js';
 
+import { createEmbedding, vectorToString } from './contextEmbeddings.js';
+
 const sql = loadSqlEquiv(import.meta.url);
 
 const SubmissionVariantSchema = z.object({
   variant: VariantSchema,
   submission: SubmissionSchema,
 });
 const GPTGradeSchema = z.object({ grade: z.number(), feedback: z.string() });
+const GradedExampleSchema = z.object({
+  submission_text: z.string(),
+  score_perc: z.number(),
+  instance_question_id: z.string(),
+});
+type GradedExample = z.infer<typeof GradedExampleSchema>;
+
+async function generateGPTPromptWithExamples({
+  question_prompt,
+  student_answer,
+  example_submissions,
+}: {
+  question_prompt: string;
+  student_answer: string;
+  example_submissions: GradedExample[];
+}): Promise<
+  {
+    role: 'system' | 'user';
+    content: string;
+  }[]
+> {
+  const messages: {
+    role: 'system' | 'user';
+    content: string;
+  }[] = [];
+
+  // Instructions for grading
+  messages.push({
+    role: 'system',
+    content:
+      'You are an instructor for a course, and you are grading assignments. You should always return the grade using a json object of 2 parameters: grade and feedback. The grade should be an integer between 0 and 100. 0 being the lowest and 100 being the highest, and the feedback should be why you give this grade, or how to improve the answer. You can say correct or leave blank when the grade is close to 100. I will provide some example answers and their corresponding grades.',
+  });
+
+  // Question prompt
+  messages.push({
+    role: 'user',
+    content: `Question: \n${question_prompt}`,
+  });
+
+  // Examples
+  for (const example of example_submissions) {
+    messages.push({
+      role: 'user',
+      content: `Example answer: \n${example.submission_text} \nGrade to this example answer: \n${example.score_perc}`,
+    });
+  }
+
+  // Student answer
+  messages.push({
+    role: 'user',
+    content: `Answer: \n${student_answer} \nHow would you grade this? Please return the json object.`,
+  });
+  return messages;
+}
+
+async function generateSubmissionEmbeddings({
+  course,
+  question,
+  assessment_question,
+  urlPrefix,
+  openai,
+}: {
+  question: Question;
+  course: Course;
+  assessment_question: AssessmentQuestion;
+  urlPrefix: string;
+  openai: OpenAI;
+}): Promise<number> {
+  const question_course = await getQuestionCourse(question, course);
+
+  const result = await queryRows(
+    sql.select_all_instance_questions,
+    {
+      assessment_id: assessment_question.assessment_id,
+      assessment_question_id: assessment_question.id,
+    },
+    InstanceQuestionSchema,
+  );
+
+  let newEmbeddingsCount = 0;
+
+  for (const instance_question of result) {
+    const { variant, submission } = await queryRow(
+      sql.select_last_variant_and_submission,
+      { instance_question_id: instance_question.id },
+      SubmissionVariantSchema,
+    );
```

Comment:
I won't block on this, but: in the best case where an embedding already exists, you're fetching an entire variant+submission only to immediately throw it away. For some hypothetical course with 1000 students each of which uploaded a 2MB PDF submission, you're pushing 2GB over the wire that you will never do _anything_ with. I'd like for this to be addressed before this gets actual production usage.

Reply:
What about this (still trying to think about better plans): what if I use a simplified version of the query that only returns the submission id here instead of the entire submission and variant information? Realistically I only need a submission id to find an embedding in the table. I can run this to get the submission later if I need to generate the embedding.

Reply:
That's exactly what I'd do here. That is: one query to retrieve just the latest submission ID, and one query that returns the submission+variant rows for a particular submission ID.

Reply:
I've updated this part